docs/ethernet.md

```diff
@@ -1,28 +1,118 @@
+endpoint is shutdown.
+
+The active ports may be queried by connecting to a Unix domain socket on the Mac
+or a named pipe on Windows and receiving diagnostic data in a Unix tar formatted
+stream.
+
+
+## Example: connection setup
+
+Consider what happens when an application inside a container in the Linux VM
+tries to make a TCP connection:
+
+- the application calls `connect`
+- the Linux kernel emits a `TCP` packet with the `SYN` flag set
+- the Linux kernel applies the `iptables` rules and consults
+  the routing table to select the outgoing interface and then transmits the frame
+- the frame is relayed to the host
+  - on windows: the interface was a `tap` device created by the `tap-vsockd`
+    process. This process reads the frame from the associated file descriptor,
+    encapsulates it and writes it to the Hyper-V socket connected to `vpnkit`.
+  - on Mac: the interface was a `virtio-net` NIC. The network driver in the VM pushes
+    the packet to a queue in memory shared with the hypervisor, the `virtio-vpnkit`
+    virtual hardware pops the packet from the queue and then writes it down a
+    Unix domain socket connected to `vpnkit`.
+- the frame is received by `vpnkit` and input into the ethernet switch.
+  - if the destination IP is not recognised: `vpnkit` creates a TCP/IP endpoint
+    using [Mirage](https://mirage.io/) [TCP/IP stack](https://github.com/mirage/mirage-tcpip)
+    with the destination IP address and configures the switch to send future
+    packets with this destination IP to this endpoint
+  - if the destination IP is recognised: the internal switch inputs the frame
+    into the TCP/IP endpoint
+- the TCP/IP endpoint observes the `SYN` flag is set and so it calls the regular
+  `connect` API to establish a regular `SOCK_STREAM` connection to that destination.
+  - if the `connect` succeeds: the TCP/IP endpoint sends back a packet with the
+    `SYN` and `ACK` flags set and the handshake continues
+  - if the `connect` fails: the TCP/IP endpoint sends back a packet with the
+    `RST` flag set to reject the connection.
+
+If all has gone well, the VM now has a TCP connection over a virtual point-to-point
+ethernet link connected to `vpnkit`, and `vpnkit` has a socket connection to
+the true destination. `vpnkit` will now proxy the data in both directions.
+
+Note that from the host kernel's point of view, there is no network connection
+to the VM and no set of associated firewall rules or routing tables. All outgoing
+connections originate from the `vpnkit` process. If the user installs some
+advanced networking or VPN software which reconfigures the routing table or
+firewall rules, it will not break the connection between `vpnkit` and the VM.
 
-The active ports may be queried over the 9P debug interface via the `ports`
-subdirectory.
+This technique for forwarding network connections is commonly known as
+[Slirp](https://en.wikipedia.org/wiki/Slirp).
```

Comment:
The wikipedia page says:
>  Even though the Slirp-derived code was heavily criticized,[11] to the date there is no competing implementation available.

Maybe someone can try to update that page? :p /cc @amirmc